lib/src/default_index/revset_engine.rs

```diff
@@ -163,6 +165,76 @@ impl<I: AsCompositeIndex> Revset for RevsetImpl<I> {
             (count, Some(count))
         }
     }
+
+    fn containing_fn(&self) -> Box<dyn Fn(&CommitId) -> bool + '_> {
+        let positions = PositionsAccumulator::new(self.index.as_composite(), self.positions());
+        Box::new(move |commit_id| positions.contains(commit_id))
+    }
+}
+
+/// Incrementally consumes positions iterator of the revset collecting
+/// positions.
+struct PositionsAccumulator<'revset, 'index> {
+    index: CompositeIndex<'index>,
+    inner: Rc<RefCell<PositionsAccumulatorInner<'revset>>>,
+}
+
+impl<'revset, 'index> PositionsAccumulator<'revset, 'index> {
+    fn new(
+        index: CompositeIndex<'index>,
+        positions_iter: Box<dyn Iterator<Item = IndexPosition> + 'revset>,
+    ) -> Self {
+        let inner = Rc::new(RefCell::new(PositionsAccumulatorInner {
+            positions_iter: Some(positions_iter),
+            consumed_positions: Vec::new(),
+        }));
+        PositionsAccumulator { index, inner }
+    }
+
+    /// Checks whether the commit is in the revset.
+    fn contains(&self, commit_id: &CommitId) -> bool {
+        let Some(position) = self.index.commit_id_to_pos(commit_id) else {
+            return false;
+        };
+
+        let mut inner = self.inner.borrow_mut();
+        if let Some(last_position) = inner.consumed_positions.last() {
+            if last_position > &position {
+                inner.consume_to(position);
+            }
+        } else {
+            inner.consume_to(position);
+        }
+
+        inner
+            .consumed_positions
+            .binary_search_by(|p| p.cmp(&position).reverse())
+            .is_ok()
+    }
+
+    #[cfg(test)]
+    fn consumed_len(&self) -> usize {
+        self.inner.borrow().consumed_positions.len()
+    }
+}
+
+/// Helper struct for [`PositionsAccumulator`] to simplify interior mutability.
```

Comment:
Can we avoid this by allowing `containing_fn()` return a `FnMut` instead of an `Fn`?

Reply:
I think `Fn` is better as a trait interface. Templater will need interior mutability to cache it, and I guess the Google backend won't need `FnMut`.

Reply:
Yeah, I'd also prefer `FnMut` but don't know what unnecessary restrictions it might impose. Let's leave it as `Fn`.